R/render-plot.R

```diff
@@ -995,3 +1027,22 @@ find_panel_ranges <- function(g, res) {
     )
   })
 }
+
+# Remember the x/y limits of discrete axes. This info is
+# necessary to properly inverse map the numeric (i.e., trained)
+# positions back to the data scale, for example:
+# https://github.com/rstudio/shiny/pull/2410#issuecomment-487783828
+# https://github.com/rstudio/shiny/pull/2410#issuecomment-488100881
+#
+# Eventually, we may want to consider storing the entire ggplot2
+# object server-side and querying information from that object
+# as we need it...that's the only way we'll ever be able to
+# faithfully brush examples like this:
+# https://github.com/rstudio/shiny/issues/2411
+add_discrete_limits <- function(domain, scale, var = "x") {
+  var <- match.arg(var, c("x", "y"))
+  if (scale$is_discrete()) {
```

Comment:
Should check how far back the `is_discrete` function is available from the scales package. If in not-too-old versions the function isn't present, then check for `is.function(scale$is_discrete)` before calling it.

Reply:
`is_discrete` is actually defined in ggplot2, and has existed since at least 1.0.1.9000

https://github.com/tidyverse/ggplot2/commit/eea7fcdbb7b6d4d5de73941359e6ad6a67d44168#diff-8516aea4a3d39cf4082ac0268ba42aefR197

Reply:
Same with `get_limits()` which seems like a safer (more official) way to about this https://github.com/tidyverse/ggplot2/blame/a92042d3f57e8500bd173131d6248e2466e015fd/R/scale-.r#L113